Changing SnackBar's dismiss direction

rickimaru18 opened the pull request:
Added new property "dismissDirection" so that users can control the SnackBar's DismissDirection type.

Fixes #76986

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

rickimaru18 committed:
Changing SnackBar's dismiss direction

Added new property "dismissDirection" so that users control the SnackBar's DismissDirection type.

Piinks reviewed (commented):
Hi @rickimaru18, welcome! And thanks for the contribution!
This change will need a test to verify that this works as expected.

rickimaru18 commented:
> Hi @rickimaru18, welcome! And thanks for the contribution!
> This change will need a test to verify that this works as expected.

Hello @Piinks, thanks for the welcome :smile: 
I just need some confirmation from @Hixie before I proceed adding the tests. Just to make sure I'm doing it correctly :smile:

rickimaru18 commented:
@Piinks 

File: snack_bar_test.dart

I would just like to know if I should consider...
1. Test for all dismiss directions? Should I combine it inside "SnackBar dismiss test" or in a separate "testWidget(...)" call?     
    // e.g. SnackBar dismiss down test, SnackBar dismiss horizontal test, etc.
2. If only one direction, adding the parameter "dismissDirection: DismissDirection.down," in "SnackBar dismiss test" would suffice? Because it will error if other engineers deleted this property.

I already have the test modification for 1 and 2. I'll commit this one when I get your response. :bow:

Piinks commented:
Hey @rickimaru18 I think you could add one test that just checks all of the directions, for each of the SnackBar constructors that are being changed here.

rickimaru18 committed:
Test SnackBar's dismiss directions

In "SnackBar dismiss test", added test for other dismiss directions:
- down
- up
- vertical
- startToEnd
- endToStart
- horizontal

Piinks commented on packages/flutter/lib/src/material/snack_bar.dart:
```diff
@@ -367,6 +368,9 @@ class SnackBar extends StatefulWidget {
   /// Called the first time that the snackbar is visible within a [Scaffold].
   final VoidCallback? onVisible;
 
+  /// Swipe gesture when dismissing this snackbar.
```
```suggestion
  /// The direction in which the SnackBar can be dismissed.
  ///
  /// Cannot be null, defaults to [DismissDirection.down].
```

Piinks commented on packages/flutter/test/material/snack_bar_test.dart:
```diff
@@ -446,39 +452,84 @@ void main() {
-    expect(find.text('bar1'), findsNothing);
-    expect(find.text('bar2'), findsOneWidget);
+
+    final Map<int, Offset> dragGestures = <int, Offset>{};
```
This is some pretty nifty test craft! Can you add some comments along the way explaining what is happening? Also, since there are two of these, we might want to roll some of the duplicated code up to be re-used.

rickimaru18 replied:
@Piinks Thanks! I already updated the test and simplified the drag gestures creation to be self-explanatory. a612ed7

rickimaru18 committed:
Test and doc comment update

- Updated SnackBar's property, "dismissDirection", doc comment.
- In SnackBar's dismiss test:
  + Created new static function "_getDragGesturesOfDismissDirections(...)" which will be used by "SnackBar dismiss test" and "SnackBar dismiss test - ScaffoldMessenger".
  + Simplify drag gestures Map creation.

rickimaru18 committed:
Comment update

Piinks reviewed (commented):
It looks like there are a bunch of failing checks here currently, can you take a look? :)

rickimaru18 commented:
@Piinks I'll try to update to latest master just like my other PR. 🙇

rickimaru18 committed:
Merge branch 'master' into snackbar_dismissible_direction

rickimaru18 committed:
Remove trailing U+0020 space character.

Piinks commented on packages/flutter/test/material/snack_bar_test.dart:
```diff
@@ -2369,3 +2409,40 @@ void main() {
+  }
+
+  return dragGestures;
+}
```
Nit: newline at end of file

rickimaru18 replied:
Committed: 2635f58

Piinks commented on packages/flutter/test/material/snack_bar_test.dart:
```diff
@@ -446,39 +452,59 @@ void main() {
-    expect(find.text('bar1'), findsNothing);
-    expect(find.text('bar2'), findsOneWidget);
+
+    final Map<DismissDirection, List<Offset>> dragGestures = _getDragGesturesOfDismissDirections(width);
+
+    for (dismissDirection in dragGestures.keys) {
+      for (final Offset dragGesture in dragGestures[dismissDirection]!) {
+        snackBarCount = 0;
+
+        expect(find.text('bar1'), findsNothing);
+        expect(find.text('bar2'), findsNothing);
+        await tester.tap(find.byKey(tapTarget)); // queue bar1
+        await tester.tap(find.byKey(tapTarget)); // queue bar2
+        expect(find.text('bar1'), findsNothing);
+        expect(find.text('bar2'), findsNothing);
+        await tester.pump(); // schedule animation for bar1
+        expect(find.text('bar1'), findsOneWidget);
+        expect(find.text('bar2'), findsNothing);
+        await tester.pump(); // begin animation
+        expect(find.text('bar1'), findsOneWidget);
+        expect(find.text('bar2'), findsNothing);
+        await tester.pump(const Duration(milliseconds: 750)); // 0.75s // animation last frame; two second timer starts here
+        await tester.drag(find.text('bar1'), dragGesture);
+        await tester.pump(); // bar1 dismissed, bar2 begins animating
+        expect(find.text('bar1'), findsNothing);
+        expect(find.text('bar2'), findsOneWidget);
+        await tester.pump(const Duration(milliseconds: 750)); // 0.75s // animation last frame; two second timer starts here
+        await tester.drag(find.text('bar2'), dragGesture);
+        await tester.pump(); // bar2 dismissed
+        expect(find.text('bar1'), findsNothing);
+        expect(find.text('bar2'), findsNothing);
+      }
+    }
```
This section of code 456-485 looks like it can be consolidated as well, it is the same as 521-550

rickimaru18 replied:
Committed: 2635f58

Piinks commented on packages/flutter/test/material/snack_bar_test.dart:
```diff
@@ -2369,3 +2409,40 @@ void main() {
+          const Offset(0.0, -50.0), // drag to top gesture
+        ];
+        break;
+      case DismissDirection.startToEnd:
+        dragGestures[val] = <Offset>[Offset(scaffoldWidth, 0.0)]; // drag to right gesture
+        break;
+      case DismissDirection.endToStart:
```
Does this account for `TextDirection`? It may be worth adding a test that used `TextDirection.ltr` versus `TextDirection.rtl` to ensure the behavior is correct.

rickimaru18 replied:
If it's to test the `Dismissible` widget, I think it is already considered in `dismissible_test.dart`.

Piinks reviewed (commented):
Hey @rickimaru18 I apologize it has taken me a few days to circle back to this. Thank you for updating the tests. :)

rickimaru18 committed:
Added common tester function for tests "SnackBar dismiss test".

rickimaru18 committed:
Merge branch 'master' into snackbar_dismissible_direction

Piinks commented:
Disregard the G testing check, confirmed it was a flake.

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/117732724-f99dd700-b1b5-11eb-8fcb-0300a1039c65.png)
:tada: Thank you for the contribution!
